src/persistence/sql.rs

```diff
@@ -1,37 +1,1008 @@
+                )
+                .bind(withdrawal.delegator.to_string())
+                .bind(withdrawal.node.to_string())
+                .fetch_one(&mut **tx)
+                .await?;
+
+                let current_amount = U256::from_str(&amount).unwrap_or(U256::ZERO);
+                let new_amount = current_amount.checked_sub(withdrawal.amount).ok_or_else(
+                    || {
+                        anyhow::anyhow!(
+                            "Underflow: withdrawal {} exceeds delegation amount {current_amount}",
+                            withdrawal.amount
+                        )
+                    },
+                )?;
+
+                // Update delegation: decrement amount and set unlocks_at and withdrawal_amount
+                let result = sqlx::query(
+                    "UPDATE delegation
+                     SET amount = $1,
+                         unlocks_at = $2,
+                         withdrawal_amount = $3
+                     WHERE delegator = $4 AND node = $5",
+                )
+                .bind(new_amount.to_string())
+                .bind(withdrawal.available_time as i64)
+                .bind(withdrawal.amount.to_string())
+                .bind(withdrawal.delegator.to_string())
+                .bind(withdrawal.node.to_string())
+                .execute(&mut **tx)
+                .await?;
+
+                if result.rows_affected() != 1 {
+                    return Err(anyhow::anyhow!(
+                        "Expected to update 1 delegation row, but {} were affected",
+                        result.rows_affected()
+                    ))
+                    .map_err(Into::into);
+                }
+            }
+            WalletDiff::NodeExited(withdrawal) => {
+                // Mark all delegations to this node as exits
+                let result = sqlx::query(
+                    "UPDATE delegation
+                     SET unlocks_at = $1, withdrawal_amount = '0'
+                     WHERE delegator = $2 AND node = $3",
+                )
+                .bind(withdrawal.available_time as i64)
+                .bind(withdrawal.delegator.to_string())
+                .bind(withdrawal.node.to_string())
+                .execute(&mut **tx)
+                .await?;
+
+                if result.rows_affected() != 1 {
+                    return Err(anyhow::anyhow!(
+                        "Expected to update 1 delegation row, but {} were affected",
+                        result.rows_affected()
+                    ))
+                    .map_err(Into::into);
+                }
+            }
+            WalletDiff::UndelegationWithdrawal(withdrawal) => {
+                let result = sqlx::query(
+                    "UPDATE delegation
+                         SET unlocks_at = 0,
+                             withdrawal_amount = '0'
+                         WHERE delegator = $1 AND node = $2",
```

Comment:
Could do a `RETURNING amount` here and if the `amount` is also 0, then we can delete the entire row.